thriftbp/server.go

```diff
@@ -91,38 +96,232 @@ func NewServer(cfg ServerConfig) (*thrift.TSimpleServer, error) {
 	return server, nil
 }
 
+// NewServerFromOptions creates a new ServerConfig instance and delegates
+// to NewServer so that it can instantiate a new thrift.TSimpleServer.
+// This is a low level api and should only be used at the recommendation
+// of the baseplate maintainers. Most application implementations should
+// rely on NewBaseplateServer.
+func NewServerFromOptions(opts ...ServerOption) (*thrift.TSimpleServer, error) {
+	cfg, err := buildConfig(opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build server config: %w", err)
+	}
+	return NewServer(cfg)
+}
+
 // NewBaseplateServer returns a new Thrift implementation of a Baseplate
-// server with the given config.
+// server with the given config. This is the primary constructor that
+// should be used when building a baseplate server.
 func NewBaseplateServer(
 	bp baseplate.Baseplate,
 	cfg ServerConfig,
 ) (baseplate.Server, error) {
-	middlewares := BaseplateDefaultProcessorMiddlewares(
-		DefaultProcessorMiddlewaresArgs{
-			EdgeContextImpl:                    bp.EdgeContextImpl(),
-			ErrorSpanSuppressor:                cfg.ErrorSpanSuppressor,
-			ReportPayloadSizeMetricsSampleRate: cfg.ReportPayloadSizeMetricsSampleRate,
-		},
+	options := append(
+		[]ServerOption{withConfigFrom(cfg)},
+		DefaultServerOptions(bp)...,
 	)
-	middlewares = append(middlewares, cfg.Middlewares...)
-	cfg.Middlewares = middlewares
-	if cfg.Logger == nil {
-		cfg.Logger = log.ZapWrapper(log.ZapWrapperArgs{
-			Level: bp.GetConfig().Log.Level,
-			KVPairs: map[string]interface{}{
-				"from": "thrift",
-			},
-		}).ToThriftLogger()
-	}
-	cfg.Addr = bp.GetConfig().Addr
-	cfg.Socket = nil
-	srv, err := NewServer(cfg)
+	return NewBaseplateServerFromOptions(bp, options...)
+}
+
+// NewBaseplateServerFromOptions returns a new Thrift implementation of a Baseplate
+// server using a config built from the supplied opts ServerOption. This is a low
+// level api and should  only be used at the recommendation of the baseplate maintainers.
+// Most application implementations should rely on NewBaseplateServer.
+func NewBaseplateServerFromOptions(
+	bp baseplate.Baseplate,
+	opts ...ServerOption,
+) (baseplate.Server, error) {
+	srv, err := NewServerFromOptions(opts...)
 	if err != nil {
 		return nil, err
 	}
 	return ApplyBaseplate(bp, srv), nil
 }
 
+// ServerOption is a type used for defining configuration arguments
+// needed for creating a baseplate server. It allows returning
+// an error to signal that a configuration value was invalid
+// or failed to be set.
+type ServerOption func(cfg *ServerConfig) error
+
+// buildConfig creates a new ServerConfig instance and applies
+// all the supplied configuration options to it.
+func buildConfig(options ...ServerOption) (ServerConfig, error) {
+	cfg := &ServerConfig{}
```

Comment:
my preference would be to use `var cfg ServerConfig` here, use `option(&cfg)` on line 152 and `return cfg, nil` on line 158.